lib/uaa/token_issuer.rb

```diff
@@ -53,9 +54,28 @@ class TokenIssuer
   include Http
 
   private
+  @client_auth_method = 'client_secret_basic'
 
   def random_state; SecureRandom.hex end
 
+  # Generates a random verifier for PKCE usage
+  def code_verifier
+    if @code_verifier
+      @verifier = @code_verifier
+    else
+      @verifier ||= SecureRandom.base64(96).tr("+/", "-_").tr("=", "")
+    end
+  end
+
+  # Calculates the challenge from code_verifier
+  def code_challenge
+    if @code_verifier
```

Comment:
Is this check necessary? Wouldn't the method `code_verifier` also perform this check?

Reply:
this is needed because you can add your own code_verifier. The code_verifier is the secret.
cf-uaa-lib needs to have this because if you have async. frameworks, then you cannot always calculate the secret on your own.
Added this for uaac because here it is needed that we shared the code verifier between local console and the extra method which creates the server for code exchange. uaac uses here different instances of cf-uaa-lib

Reply:
Understood. I was referring to how the method `def code_verifier`:
```
  def code_verifier
    if @code_verifier
      @verifier = @code_verifier
    else
      @verifier ||= SecureRandom.base64(96).tr("+/", "-_").tr("=", "")
    end
  end
```
which is called in the else case of `def code_challenge`, would check whether there is a passed-in `@code_verifier` already. So performing the same check in `def code_challenge` doesn't seem necessary. 
That said, this is just a minor point. It's not a big deal either way.

Reply:
I am not sure if I get it now, can you please change the code if it is not that what you mean.

Reply:
I just meant that: Isn't the previous code below:
```
  def code_verifier
    if @code_verifier
      @verifier = @code_verifier
    else
      @verifier ||= SecureRandom.base64(96).tr("+/", "-_").tr("=", "")

  def code_challenge
    if @code_verifier
      @challenge = Digest::SHA256.base64digest(@code_verifier).tr("+/", "-_").tr("=", "")
    else
      @challenge ||= Digest::SHA256.base64digest(code_verifier).tr("+/", "-_").tr("=", "")
    end
  end
```
be equivalent to below?
```
  def code_verifier
    if @code_verifier
      @verifier = @code_verifier
    else
      @verifier ||= SecureRandom.base64(96).tr("+/", "-_").tr("=", "")

  def code_challenge
      @challenge ||= Digest::SHA256.base64digest(code_verifier).tr("+/", "-_").tr("=", "")
  end
```

Since the `if @code_verifier` check will be performed by `def code_verifier`?

Reply:
Your most recent update of the code:
```
  # Generates a random verifier for PKCE usage
  def code_verifier
    if not @code_verifier.nil?
      @verifier = @code_verifier
    else
      @verifier ||= SecureRandom.base64(96).tr("+/", "-_").tr("=", "")

  # Calculates the challenge from code_verifier
  def code_challenge
    @challenge = Digest::SHA256.base64digest(@code_verifier).tr("+/", "-_").tr("=", "")
  end
```
seems to change the logic though. Since now `def code_challenge` is always using `@code_verifier` rather than sometimes generating when `@code_verifier` is empty (which I thought was what you wanted before?).

Reply:
the logic is, you can bring a code (secret) or rely on generated one. Based on this challenge is calculated.
This logic should stay and with ||= the calculation is only done once , ruby keeps status.

I reduced the complexity because before I thought PKCE can be handled in cf-uaa-lib internally only, but in uaac I realized, that this is not possible, therefore maybe to much code before